std/machines/hash/keccakf16_memory.asm

```diff
@@ -0,0 +1,876 @@
+    link if first_step ~> (addr_h[20], addr_l[20]) = add_sub.add(input_addr_h, input_addr_l, 0, 84);
+    link if first_step ~> (addr_h[21], addr_l[21]) = add_sub.add(input_addr_h, input_addr_l, 0, 88);
+    link if first_step ~> (addr_h[22], addr_l[22]) = add_sub.add(input_addr_h, input_addr_l, 0, 92);
+    link if first_step ~> (addr_h[23], addr_l[23]) = add_sub.add(input_addr_h, input_addr_l, 0, 96);
+    link if first_step ~> (addr_h[24], addr_l[24]) = add_sub.add(input_addr_h, input_addr_l, 0, 100);
+    link if first_step ~> (addr_h[25], addr_l[25]) = add_sub.add(input_addr_h, input_addr_l, 0, 104);
+    link if first_step ~> (addr_h[26], addr_l[26]) = add_sub.add(input_addr_h, input_addr_l, 0, 108);
+    link if first_step ~> (addr_h[27], addr_l[27]) = add_sub.add(input_addr_h, input_addr_l, 0, 112);
+    link if first_step ~> (addr_h[28], addr_l[28]) = add_sub.add(input_addr_h, input_addr_l, 0, 116);
+    link if first_step ~> (addr_h[29], addr_l[29]) = add_sub.add(input_addr_h, input_addr_l, 0, 120);
+    link if first_step ~> (addr_h[30], addr_l[30]) = add_sub.add(input_addr_h, input_addr_l, 0, 124);
+    link if first_step ~> (addr_h[31], addr_l[31]) = add_sub.add(input_addr_h, input_addr_l, 0, 128);
+    link if first_step ~> (addr_h[32], addr_l[32]) = add_sub.add(input_addr_h, input_addr_l, 0, 132);
+    link if first_step ~> (addr_h[33], addr_l[33]) = add_sub.add(input_addr_h, input_addr_l, 0, 136);
+    link if first_step ~> (addr_h[34], addr_l[34]) = add_sub.add(input_addr_h, input_addr_l, 0, 140);
+    link if first_step ~> (addr_h[35], addr_l[35]) = add_sub.add(input_addr_h, input_addr_l, 0, 144);
+    link if first_step ~> (addr_h[36], addr_l[36]) = add_sub.add(input_addr_h, input_addr_l, 0, 148);
+    link if first_step ~> (addr_h[37], addr_l[37]) = add_sub.add(input_addr_h, input_addr_l, 0, 152);
+    link if first_step ~> (addr_h[38], addr_l[38]) = add_sub.add(input_addr_h, input_addr_l, 0, 156);
+    link if first_step ~> (addr_h[39], addr_l[39]) = add_sub.add(input_addr_h, input_addr_l, 0, 160);
+    link if first_step ~> (addr_h[40], addr_l[40]) = add_sub.add(input_addr_h, input_addr_l, 0, 164);
+    link if first_step ~> (addr_h[41], addr_l[41]) = add_sub.add(input_addr_h, input_addr_l, 0, 168);
+    link if first_step ~> (addr_h[42], addr_l[42]) = add_sub.add(input_addr_h, input_addr_l, 0, 172);
+    link if first_step ~> (addr_h[43], addr_l[43]) = add_sub.add(input_addr_h, input_addr_l, 0, 176);
+    link if first_step ~> (addr_h[44], addr_l[44]) = add_sub.add(input_addr_h, input_addr_l, 0, 180);
+    link if first_step ~> (addr_h[45], addr_l[45]) = add_sub.add(input_addr_h, input_addr_l, 0, 184);
+    link if first_step ~> (addr_h[46], addr_l[46]) = add_sub.add(input_addr_h, input_addr_l, 0, 188);
+    link if first_step ~> (addr_h[47], addr_l[47]) = add_sub.add(input_addr_h, input_addr_l, 0, 192);
+    link if first_step ~> (addr_h[48], addr_l[48]) = add_sub.add(input_addr_h, input_addr_l, 0, 196);
```

Comment:
this looks expensive haha
fine to merge for now, but we should use the add4 trick

Reply:
Ok. I'm opening a new PR soon.